### PR TITLE
[SPARK-25150][SQL] Rewrite condition when deduplicate Join

### PR DESCRIPTION
**Pull request by peter-toth**
## What changes were proposed in this pull request?

ResolveReferences rule modifies (deduplicates) conflicting AttributeReferences in left and right side of a Join. It changes the right side to avoid conflicts but it forgot to change the references in the condition of the join. Solution is to rewrite the condition too.

Details:

With this PR, if any of the resolved attributes in a join condition refer to an attribute that is replaced during the deduplication (`dedupRight()`) then the join condition is modified to reflect the change.
The modification is done regardless whether the reference is on the left or the right side of the condition.

This PR has no effect on unresolved attributes. Those attributes are resolved later by different rules. 

This PR helps in those cases where same origin dataframes (`b`, `c`) are joined to a different origin one (`a`): 
```
val a = spark.range(1, 5)
val b = spark.range(10)
val c = b.filter($"id" % 2 === 0)
val r = a.join(b, a("id") === b("id"), "inner").join(c, a("id") === c("id"), "inner")
```
Here the result of `a join b` contains an `id` attribute coming from `b`. That attribute conflicts with the `id` attribute of `c` in the second join so `c("id")` is deduplicated to let's say `X` and the join condition `a("id") === c("id")` needs to be modified to `a("id") === X` to get correct results.

I believe it is also worth explaining some situations where this PR does't help (but also does no harm). In cases where same origin dataframes are joined to each other:
```
val b = spark.range(10)
val c = b.filter($"id" % 2 === 0)
val r = b.join(c, b("id") === c("id"), "inner")
```
In this latter case the deduplication is applied as both `b` and `c` have an `id` attribute (that actually refer to the same resolved attribute (have the same `ExprId`)). Deduplication results `c("id")` to be changed to let's say `X`. So in this case both `b("id")` and `c("id")` in the join condition are modified to `X` by the changes introduced in this PR. The reason why the result will be correct with this PR (and is correct without this PR) is that the current implementation of Spark contains a "hack" in `Dataframe` that fixes `EqualTo` and `EqualNullSafe` type of conditions with identical references on both sides. The hack rewrites these join conditions by re-resolving the attributes with the name of the attribute on both the left and right sides.

Please note that until a resolved attribute doesn't contain some kind of reference to it's dataframe not all  cases of join can be handled properly. Such an initiative can be found in https://github.com/apache/spark/pull/21449.

## How was this patch tested?

Added unit test.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Review by mgaido91 — COMMENTED** [review]
This change makes sense to me, despite I quite don't get how this didn't come out earlier.

@peter-toth may you please improve the PR description as suggested by @maropu? May you please also check whether this is a regression from 2.2? In that case, it might be a blocker for the next 2.3 release.

cc @cloud-fan @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala [hunk 1]
```diff
@@ -295,4 +295,17 @@ class DataFrameJoinSuite extends QueryTest with SharedSQLContext {
...
+      val a = spark.range(1, 5)
+      val b = spark.range(10)
+      val c = b.filter($"id" % 2 === 0)
+
+      val r = a.join(b, a("id") === b("id"), "inner").join(c, a("id") === c("id"), "inner")
+
+      checkAnswer(r, Row(2, 2, 2) :: Row(4, 4, 4) :: Nil)
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala [hunk 2]
```diff
@@ -295,4 +295,17 @@ class DataFrameJoinSuite extends QueryTest with SharedSQLContext {
       df.join(df, df("id") <=> df("id")).queryExecution.optimizedPlan
     }
   }
+
+  test("SPARK-25150: Attribute deduplication handles attributes in join condition properly") {
+    withSQLConf(SQLConf.CROSS_JOINS_ENABLED.key -> "false") {
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala]: actually we can remove this

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -895,6 +897,13 @@ class Analyzer(
       case _ => e.mapChildren(resolve(_, q))
     }
 
+    private def rewriteJoinCondition(
+        e: Expression,
+        attributeRewrites: AttributeMap[Attribute]): Expression = e match {
+      case a: Attribute => attributeRewrites.getOrElse(a, a)
+      case _ => e.mapChildren(rewriteJoinCondition(_, attributeRewrites))
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: sorry, we can't do that, it wouldn't mean the same

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala [hunk 4]
```diff
@@ -295,4 +295,17 @@ class DataFrameJoinSuite extends QueryTest with SharedSQLContext {
...
+      val a = spark.range(1, 5)
+      val b = spark.range(10)
+      val c = b.filter($"id" % 2 === 0)
+
+      val r = a.join(b, a("id") === b("id"), "inner").join(c, a("id") === c("id"), "inner")
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala]: I think we do need `a` here.
If we dropped `a` and the test would become like:
```scala
    val b = spark.range(1, 5)
    val c = b.filter($"id" % 2 === 0)
    val r = b.join(c, b("id") === c("id"), "inner")

    checkAnswer(r, Row(2, 2) :: Row(4, 4) :: Nil)
```
then the test would pass even without the fix. This is because we have a special case to handle {{id = id}} like conditions in case of EqualTo and EqualNullSafe in Dataset.

My fix comes into play in some other cases where there is an AttributeReference change in the right side of a join due to deduplication.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: oh, sorry, I see now, sorry.

**Comment by cloud-fan** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -805,10 +807,10 @@ class Analyzer(
            * that this rule cannot handle. When that is the case, there must be another rule
            * that resolves these conflicts. Otherwise, the analysis will fail.
            */
-          right
+          (right, AttributeMap.empty[Attribute])
         case Some((oldRelation, newRelation)) =>
           val attributeRewrites = AttributeMap(oldRelation.output.zip(newRelation.output))
-          right transformUp {
+          (right transformUp {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: `(right transformUp {` -> `val newRight = right transformUp {`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -817,7 +819,7 @@ class Analyzer(
               case s: SubqueryExpression =>
                 s.withNewPlan(dedupOuterReferencesInSubquery(s.plan, attributeRewrites))
             }
-          }
+          }, attributeRewrites)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Then, `(newRight, attributeRewrites)`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -805,10 +807,10 @@ class Analyzer(
            * that this rule cannot handle. When that is the case, there must be another rule
            * that resolves these conflicts. Otherwise, the analysis will fail.
            */
-          right
+          (right, AttributeMap.empty[Attribute])
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Since I don't want to build an empty object per call, how about defining an empty map in this class field? e.g.,
```
  object ResolveReferences extends Rule[LogicalPlan] {

    private val emptyAttrMap = new AttributeMap[Attribute](Map.empty)
```

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: just commented the same meanwhile :) sorry, I saw your comments only now :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -921,12 +930,16 @@ class Analyzer(
         failAnalysis("Invalid usage of '*' in explode/json_tuple/UDTF")
 
       // To resolve duplicate expression IDs for Join and Intersect
-      case j @ Join(left, right, _, _) if !j.duplicateResolved =>
-        j.copy(right = dedupRight(left, right))
+      case j @ Join(left, right, _, condition) if !j.duplicateResolved =>
+        val (dedupedRight, attributeRewrites) = dedupRight(left, right)
+        val changedCondition = condition.map(rewriteJoinCondition(_, attributeRewrites))
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: thanks! fixed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 5 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 6 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: done

**Comment by peter-toth** [comment]
Also added missing `if attr.resolved` which I think will fix the UT issues.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala [hunk 9]
```diff
@@ -295,4 +295,14 @@ class DataFrameJoinSuite extends QueryTest with SharedSQLContext {
       df.join(df, df("id") <=> df("id")).queryExecution.optimizedPlan
     }
   }
+
+  test("SPARK-25150: Attribute deduplication handles attributes in join condition properly") {
+    val a = spark.range(1, 5)
+    val b = spark.range(10)
+    val c = b.filter($"id" % 2 === 0)
+
+    val r = a.join(b, a("id") === b("id"), "inner").join(c, a("id") === c("id"), "inner")
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala]: That simpler join doesn't hit the issue. It is handled by a different rule `ResolveNaturalAndUsingJoin`.

**Comment by peter-toth** [comment]
@mgaido91 , 2.2 also suffered from this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 10]
```diff
@@ -754,11 +754,16 @@ class Analyzer(
    * a logical plan node's children.
    */
   object ResolveReferences extends Rule[LogicalPlan] {
+
+    private val emptyAttrMap = new AttributeMap[Attribute](Map.empty)
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I'd prefer to do [what I suggested previously](https://github.com/apache/spark/pull/22318#discussion_r214678100) as it would be easier to reuse if it will be needed. In this way, next time we need again an empty AttributeMap we need to create a new one and we'd end up with several of them.

cc @cloud-fan @maropu WDYT?

**Review by mgaido91 — COMMENTED** [review]
@peter-toth may you please also update the title of the PR to be more precise than "FIx"? Like "Rewrite condition when deduplicate Join" or something similar which explains better what the PR does?

Just left one comment, otherwise LGTM, thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 10 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: @mgaido91 , I agree with you and happy to do it, just saw your concerns about binary compatibility.
@cloud-fan  @maropu please share your thoughts on this.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: its ok to me.

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala [hunk 11]
```diff
@@ -23,12 +23,14 @@ package org.apache.spark.sql.catalyst.expressions
  * of the name, or the expected nullability).
  */
 object AttributeMap {
+  var empty = new AttributeMap(Map.empty)
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala]: `var` -> `val`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 12]
```diff
@@ -754,11 +754,14 @@ class Analyzer(
    * a logical plan node's children.
    */
   object ResolveReferences extends Rule[LogicalPlan] {
+
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: unneeded newline

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala (hunk 11 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala]: oh thanks, fixed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 12 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: fixed

**Comment by mgaido91** [comment]
LGTM

**Comment by kiszk** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala [hunk 13]
```diff
@@ -23,12 +23,14 @@ package org.apache.spark.sql.catalyst.expressions
  * of the name, or the expected nullability).
  */
 object AttributeMap {
+  val empty = new AttributeMap(Map.empty)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala]:
```Scala
def empty[A]: AttributeMap[A] = new AttributeMap(Map.empty)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala [hunk 14]
```diff
@@ -23,12 +23,14 @@ package org.apache.spark.sql.catalyst.expressions
...
+
   def apply[A](kvs: Seq[(Attribute, A)]): AttributeMap[A] = {
     new AttributeMap(kvs.map(kv => (kv._1.exprId, kv)).toMap)
   }
 }
 
-class AttributeMap[A](val baseMap: Map[ExprId, (Attribute, A)])
+class AttributeMap[+A](val baseMap: Map[ExprId, (Attribute, A)])
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala]: revert the change here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 15]
```diff
@@ -805,10 +807,10 @@ class Analyzer(
            * that this rule cannot handle. When that is the case, there must be another rule
            * that resolves these conflicts. Otherwise, the analysis will fail.
            */
-          right
+          (right, AttributeMap.empty)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: `AttributeMap.empty[Attribute]`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 16]
```diff
@@ -921,12 +924,18 @@ class Analyzer(
         failAnalysis("Invalid usage of '*' in explode/json_tuple/UDTF")
 
       // To resolve duplicate expression IDs for Join and Intersect
-      case j @ Join(left, right, _, _) if !j.duplicateResolved =>
-        j.copy(right = dedupRight(left, right))
+      case j @ Join(left, right, _, condition) if !j.duplicateResolved =>
+        val (dedupedRight, attributeRewrites) = dedupRight(left, right)
+        val changedCondition = condition.map(_.transform {
+          case attr: Attribute if attr.resolved => attributeRewrites.getOrElse(attr, attr)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: This looks fishy. We will not rewrite attributes in the join condition when they are not resolved? What will happen then?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala (hunk 13 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala]: @gatorsmile this was the initial version. then came the idea of @mgaido91 and @maropu to use val (instead of def) either here or in ResolveReferences to spare some empty() calls

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 16 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Simply they will be resolved later.
`attributeRewrites` returned by `dedupRight()` can contain resolved attributes only.

Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: but actually there might be an issue here, instead of `attributeRewrites.getOrElse(attr, attr)` I might need to use `dedupAttr(attr, attributeRewrites)` which retains attribute qualifier

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala (hunk 13 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala]: Reverted to the previous one where we have one val in ResolveReferences.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 15 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: ok, reverted

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala (hunk 14 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeMap.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 16 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: changed to `dedupAttr()`

**Comment by mgaido91** [comment]
retest this please

**Comment by maropu** [comment]
LGTM

**Comment by cloud-fan** [comment]
How does this work? When we have duplicated attributes in the join condition, how can we know which attribute comes from which side?

**Comment by cloud-fan** [comment]
Can you define the scope of this PR? In which case we should change the references in the join condition?

**Comment by peter-toth** [comment]
@cloud-fan , I added some explanation to the description in which cases this PR helps and also where it doesn't.

**Comment by peter-toth** [comment]
@cloud-fan,  does the new description defines the scope as you suggested? Is there anything I can add to this PR?

**Comment by peter-toth** [comment]
@cloud-fan could you please help me with this PR and take it one step forward?

**Comment by viirya** [comment]
For a query similar to the one in the PR description:

`a.join(b, a("id") === b("id"), "inner").join(c, a("id") === b("id"), "inner"`

It is interpreted as a cartesian products now. But after this change, it is as the same as the query: 

`a.join(b, a("id") === b("id"), "inner").join(c, a("id") === c("id"), "inner")`

**Comment by peter-toth** [comment]
Thanks @viirya, your analysis is correct.

Unfortunately an attribute doesn't have a reference to its dataset so I don't think this scenario can be solved easily. I believe the good solution would be something like https://github.com/apache/spark/pull/21449 

But I would argue that my PR still does make sense and an `a("id") === b("id")` condition in a join where `c` is joined is not expected, and actually very likely a typo.

**Comment by peter-toth** [comment]
Also please consider that currently (and also after this PR) using `b` and `c` from the description: 
```
b.join(c, b("id") === b("id"), "inner")
```
is interpreted as
```
b.join(c, b("id") === c("id"), "inner")
```
so my PR just extends this feature to the case with subsequent joins.

@cloud-fan , @viirya what do you think?

**Comment by peter-toth** [comment]
@srowen, I saw your last comment on https://github.com/peter-toth/spark/tree/SPARK-25150. I submitted this PR to solve that ticket and I believe the description here explains what is the real issue there.
I would appreciate your thoughts on this PR, unfortunately it got stuck a bit lately. Thanks.

**Comment by mgaido91** [comment]
@srowen the change seems fine to me as I think this does improve the behavior of attributes deduplication (I think it was a bug not rewriting the join condition with the new references).

The point is that in order to have a really correct and expected behavior in all the conditions we need to keep a reference of the dataset an attribute is from. That is why I created #21449 and I know @cloud-fan mentioned he had a similar PR for the same reason. But that is another story. In general I think both this and something like #21449 is needed in order to handle properly all the cases.

**Comment by maropu** [comment]
@peter-toth we cannot fix the issue of the description without changing [the existing behaviour](https://github.com/apache/spark/pull/22318#issuecomment-426317617)?

**Comment by peter-toth** [comment]
@maropu, I don't think we can. Actually this is how we deal with [simpler joins](https://github.com/apache/spark/pull/22318#issuecomment-427080091)
Do you think changing the behaviour is unacceptable?

**Comment by maropu** [comment]
In the example @viirya described above (https://github.com/apache/spark/pull/22318#issuecomment-426317617), I think the interpretation is unclear to most users and I'm fairly concerned that it could be error-prone...
